Hack week/stats widget - Step 1: Display current day v4 stats, if available, for the current site

anitaa1990 opened the pull request:
This PR completes the first step of the Stats widget Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/2041).

#### Changes
- Adds a new widget to the app that displays the current day stats, if v4 stats is available for the site.
- An error message is displayed in the widget if the user is not logged in to the app.
- An error message is displayed in the widget if the store does not have WC-Admin installed or if WooCommerce 4.0 is not installed.

#### Screenshot
##### Adding a new widget for site
<img width="300" src="https://user-images.githubusercontent.com/22608780/76514979-d034ce80-647e-11ea-9ba6-23d8e2cfd553.png">

##### Adding a new widget when user is not logged in
<img width="300" src="https://user-images.githubusercontent.com/22608780/76515357-741e7a00-647f-11ea-9ae2-7ff87ff9565b.png">

##### Adding a new widget when v4 stats is not available for a site 
<img width="300" src="https://user-images.githubusercontent.com/22608780/76515684-0888dc80-6480-11ea-94a8-2ced052ed73c.png">


#### Testing
- Run `TodayWidgetListViewModelTest` unit test.
- Add widget to a site that has WC-admin or Woo 4.0 available:
   - Login to a store that has WooCommerce 4.0 or WC-admin installed.
   - Long-press `home screen`.
   - Tap `Widgets` option.
   - Scroll to ***Woo*** widget.
   - Long press the ***Today*** widget and place on the home screen.
   - Notice the current day stats (visitors, order count and revenue) is shown in widget.
   - Click on the widget and notice that you are redirected to the My Store `Today` TAB.
- Verify that the widget data is refreshed when store is changed.
   - Complete the previous test and add a widget to the app.
   - Open the app and switch to a different store.
   - Close the app and notice that the stats data in the widget is changed to reflect the site selected in the previous step.
- Verify that error message is displayed when user is not logged in.
   - Complete the previous test and add a widget to the app.
   - Open the app and logout of the app.
   - Close the app and notice that an error is displayed in the widget when you are logged out.
- Verify that error message is displayed when stats v4 is not available for a site:
   - Complete the first test and add a widget to the app.
   - Open the `wp-admin` of the current site you are logged into.
   - Disable WC-Admin plugin for the site and refresh the app.
   - Notice the stats reverted banner is displayed in the app.
   - Close the app and notice that the stats widget displays an error message.

#### Notes
- Note that is PR is to be merged to a feature branch and will not be available for end users until all the steps in the [master thread](https://github.com/woocommerce/woocommerce-android/issues/2041) is completed.
- This PR only adds the logic to display the stats widget for the current site. The option to support multiple sites will take place in a subsequent PR.
- There are also some TODO's to add tracking events, which will again take place in a separate PR. 

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Added a wrapper class for AppPrefs to make unit test more easier

anitaa1990 committed:
Added ViewModel class to fetch current day stats for widget

anitaa1990 committed:
Added unit tests for the current day stats widget ViewModel

anitaa1990 committed:
Added provider + service class for current day stats widget

anitaa1990 committed:
Added UI for displaying stats widget

anitaa1990 committed:
Added layout file to ViewModel for displaying stats widget

anitaa1990 committed:
Added utils class for stats widget

anitaa1990 committed:
Design changes for stats widget

anitaa1990 committed:
Added config for today stats widget

anitaa1990 committed:
Fixed order of widgets items displayed in stats widget

anitaa1990 committed:
Added logic to update widget data when widget is added/deleted/updated

anitaa1990 committed:
Added logic to update widget data when stats is refreshed

anitaa1990 committed:
Added logic to update widget data when user logs out of the app

anitaa1990 committed:
Design changes to stats widget

anitaa1990 committed:
Fixed bug where visitor stats was always zero

anitaa1990 committed:
Removed redundant param from viewmodel & fixes failing unit tests

anitaa1990 committed:
Fixed lint issue

anitaa1990 committed:
Added code comments to stats widget

anitaa1990 committed:
Added preview image to widget

anitaa1990 committed:
Added logic to display widget title when error message is displayed

Garance91540 commented:
Looking awesome! Quick note, for the "Adding a new widget when user is not logged in", there's a small typo it should be "WooCommerce" not "Woocommerce"

anitaa1990 committed:
Fix typo in strings.xml

anitaa1990 commented:
Thanks for the quick review @Garance91540! I've fixed the typo in  59c96ac. This PR is ready for review 👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/widgets/WidgetUpdater.kt:
```diff
@@ -0,0 +1,70 @@
+    class StatsWidgetUpdaters
+    @Inject constructor(
+        private val todayWidgetUpdater: TodayWidgetUpdater,
+        private val appPrefsWrapper: AppPrefsWrapper,
```
It looks like `appPrefsWrapper` is unused. Does it need to be here?

anitaa1990 replied:
Ah, yes. I should have mentioned that in the commit. We would ideally use the `appPrefsWrapper` to fetch the siteId of the widget and update only that particular widget, instead of updating all the active widgets. It's currently not used but would be useful when supporting multiple sites for a widget.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStorePresenter.kt:
```diff
@@ -190,6 +193,7 @@ class MyStorePresenter @Inject constructor(
                 val visitorStats = wcStatsStore.getNewVisitorStats(
                         selectedSite.get(), event.granularity, event.quantity, event.date, event.isCustomField
                 )
+                statsWidgetUpdaters.updateTodayWidget()
```
Is there some way we can only do this if the user has enabled the widget? The reason I ask is we're keeping the updater in memory and updating it even when the widget isn't enabled.

anitaa1990 replied:
Good point! The idea is to update the widget every time the store stats refresh is successful. We do check if the widget is enabled before updating it, but that check takes place in the [WidgetUpdater](https://github.com/woocommerce/woocommerce-android/blob/5e5d38585e68ad0bf3fc1ef87b8920a563f8c146/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/widgets/WidgetUpdater.kt#L63) class. We could move that logic to this class and only call the updater if the widget is enabled but that could mean we would have to duplicate the same logic in other places in the app, should we choose to enable widgets for other features.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/widgets/stats/StatsWidget.kt:
```diff
@@ -0,0 +1,69 @@
+        if (appWidgetId > -1) {
+            widgetUpdater.updateAppWidget(
+                    context,
+                    appWidgetId = appWidgetId
```
np: `appWidgetId =` seems unnecessary given the variable has that same name.

anitaa1990 replied:
Fixed in 368870f 👍

nbradbury commented:
@anitaa1990 When I'm logged out, tapping "Retry" in the widget doesn't appear to do anything. Is that something you're adding later?

![retry](https://user-images.githubusercontent.com/3903757/76864021-ccb29600-6836-11ea-93dc-35edfe9f12a7.png)

nbradbury commented:
Really excellent work on this, @anitaa1990! I had just a few minor comments.

anitaa1990 committed:
Fixed nitpick when updating stats widget

anitaa1990 commented:
Thanks for the review @nbradbury!

> When I'm logged out, tapping "Retry" in the widget doesn't appear to do anything. Is that something you're adding later?

Nice catch! I added a separate item in this issue thread for updating the error scenarios since this was just the design I had copied from the WordPress app and it might be updated for woo in the future.

> Also, after I log back in the widget continues to show the error message and "Retry" button. It doesn't update until I tap on it. 

Hmm, I'm not able to reproduce this scenario. The widget is only updated after the stats fetch is successful though. If the app is closed after login but before stats v4 data is fetched, then it is not updated. Does this sound like the scenario you are facing?

nbradbury commented:
> The widget is only updated after the stats fetch is successful though. If the app is closed after login but before stats v4 data is fetched, then it is not updated. Does this sound like the scenario you are facing?

In this case, I'm using a site that _doesn't_ have the v4 stats enabled. As this GIF shows, the widget remains empty even after logging in and forcing stats to refresh.

![widget](https://user-images.githubusercontent.com/3903757/77064555-0f4eac80-69b6-11ea-9777-60d2732c9ba5.gif)

anitaa1990 committed:
Added logic to update widget when new stats is available

anitaa1990 commented:
> In this case, I'm using a site that doesn't have the v4 stats enabled. As this GIF shows, the widget remains empty even after logging in and forcing stats to refresh.

Ah, ok, thanks for the details. I had added the refresh only for `MyStorePresenter` since we would be removing support for the old stats UI soon. But I didn't take into account this scenario. I've fixed that in 69f0844 👍

anitaa1990 committed:
Fixed failing unit test in DashboardPresenterTest

nbradbury reviewed (approved):
:shipit:
